<launch>
  <arg name="drone_id"/>
  <arg name="drone_num"/>
  <arg name="model"/>

  <!-- Debugging -->
  <arg name="launch_prefix" default=""/> <!-- xterm -T 'Agent $(arg drone_id)' -e -->
  
  <!-- Environment size -->
  <arg name="map_size_x" default="35.0"/>
  <arg name="map_size_y" default="35.0"/>
  <arg name="map_size_z" default=" 3.5"/>
  
  <arg name="box_min_x" default="-7"/>
  <arg name="box_min_y" default="-15"/>
  <arg name="box_min_z" default="0"/> 
  <arg name="box_max_x" default="7"/>
  <arg name="box_max_y" default="15"/>
  <arg name="box_max_z" default="1.7"/>

  <arg name="odometry_topic"/>
  <arg name="sensor_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>

  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>
  <arg name="max_vel"/>
  <arg name="max_acc"/>
  <arg name="simulation" default="true"/>
  <arg name="single_expo" default="false"/>

  <!-- Ablation study -->
  <arg name="role_assigner_fixed" default="false"/>
  <arg name="role_assigner_fix_role" default="0"/> <!-- 0: EXPL, 1: GC -->
  <arg name="active_collaboration" default="true"/>
  <arg name="communication_range" default="100000.0"/>
  
  <!-- Logging -->
  <arg name="log_folder"/>

  <!-- main node -->
  <node pkg="exploration_manager" name="exploration_node_$(arg drone_id)" type="fame_node" output="log" launch-prefix="$(arg launch_prefix)">
    <remap from ="/odom_world" to="$(arg odometry_topic)"/>
    <remap from ="/map_ros/pose"   to = "$(arg sensor_pose_topic)"/> 
    <remap from ="/map_ros/depth" to = "$(arg depth_topic)"/>
    <remap from ="/map_ros/cloud" to="$(arg cloud_topic)"/>

    <remap from="/planning/replan" to="/planning/replan_$(arg drone_id)" />
    <remap from="/planning/new" to="/planning/new_$(arg drone_id)" />
    <remap from="/planning/bspline" to="/planning/bspline_$(arg drone_id)" />
    <remap from="/stop" to="/stop_$(arg drone_id)"/>
    <remap from="/heartbit" to="/heartbit_$(arg drone_id)"/>
    <remap from="/trigger_emergency" to="/trigger_emergency_$(arg drone_id)"/>

    <remap from="/swarm_expl/drone_state_send" to="/swarm_expl/drone_state" />
    <remap from="/swarm_expl/drone_state_recv" to="/swarm_expl/drone_state" />
    <remap from="/swarm_expl/pair_opt_send" to="/swarm_expl/pair_opt" />
    <remap from="/swarm_expl/pair_opt_recv" to="/swarm_expl/pair_opt" />
    <remap from="/swarm_expl/pair_opt_res_send" to="/swarm_expl/pair_opt_res" />
    <remap from="/swarm_expl/pair_opt_res_recv" to="/swarm_expl/pair_opt_res" />
    <remap from="/swarm_expl/grid_tour_send" to="/swarm_expl/grid_tour" />
    <remap from="/swarm_expl/hgrid_send" to="/swarm_expl/hgrid" />
    <remap from="/multi_map_manager/chunk_stamps_send" to="/multi_map_manager/chunk_stamps" />
    <remap from="/multi_map_manager/chunk_data_send" to="/multi_map_manager/chunk_data" />
    <remap from="/multi_map_manager/chunk_stamps_recv" to="/multi_map_manager/chunk_stamps" />
    <remap from="/multi_map_manager/chunk_data_recv" to="/multi_map_manager/chunk_data" />
    <remap from="/planning/swarm_traj_recv" to="/planning/swarm_traj" />
    <remap from="/planning/swarm_traj_send" to="/planning/swarm_traj" />

    <remap from="/planning_vis/trajectory" to="/planning_vis/trajectory_$(arg drone_id)" />
    <remap from="/planning_vis/frontier" to="/planning_vis/frontier_$(arg drone_id)" />
    <remap from="/planning_vis/viewpoints" to="/planning_vis/viewpoints_$(arg drone_id)" />
    <remap from="/planning_vis/labeled_frontier" to="/planning_vis/labeled_frontier_$(arg drone_id)" />
    <remap from="/planning_vis/role" to="/planning_vis/role_$(arg drone_id)" />
    <remap from="/planning_vis/current_goal" to="/planning_vis/current_goal_$(arg drone_id)" />
    <remap from="/planning_vis/infront_frontiers" to="/planning_vis/infront_frontiers_$(arg drone_id)" />
    <remap from="/planning_vis/trails_tour" to="/planning_vis/trails_tour_$(arg drone_id)" />

    <remap from="/sdf_map/occupancy_all" to="/sdf_map/occupancy_all_$(arg drone_id)" />
    <remap from="/sdf_map/free" to="/sdf_map/free_$(arg drone_id)" />
    <remap from="/sdf_map/occupancy_local" to="/sdf_map/occupancy_local_$(arg drone_id)" />
    <remap from="/sdf_map/occupancy_local_inflate" to="/sdf_map/occupancy_local_inflate_$(arg drone_id)" />
    <remap from="/sdf_map/unknown" to="/sdf_map/unknown_$(arg drone_id)" />
    <remap from="/sdf_map/update_range" to="/sdf_map/update_range_$(arg drone_id)" />
    <remap from="/sdf_map/basecoor" to="/swarm_sim_tf/basecoor_$(arg drone_id)" if="$(arg simulation)"/>
    <remap from="/sdf_map/basecoor" to="/swarm_drones/swarm_drone_basecoor" unless="$(arg simulation)"/>
    <remap from="/sdf_map/depth_cloud" to="/sdf_map/depth_cloud_$(arg drone_id)"/>
    

    <!--
    <param name="sdf_map/resolution"      value="$(arg map_resolution)" /> 
    <param name="sdf_map/map_size_x"   value="$(arg map_size_x)" /> 
    <param name="sdf_map/map_size_y"   value="$(arg map_size_y)" /> 
    <param name="sdf_map/map_size_z"   value="$(arg map_size_z)" /> 
    -->
    <param name="sdf_map/obstacles_inflation"     value="0.0" /> 
    <!-- <param name="sdf_map/obstacles_inflation"     value="0.0" />  -->
    <param name="sdf_map/local_bound_inflate"    value="0.5"/>
    <param name="sdf_map/local_map_margin" value="50"/>
    <param name="sdf_map/ground_height"        value="-1.0"/>
    <param name="sdf_map/default_dist"        value="0.5"/>
    
    <param name="sdf_map/p_hit"  value="0.65"/>
    <param name="sdf_map/p_miss" value="0.35"/>
    <param name="sdf_map/p_min"  value="0.12"/>
    <param name="sdf_map/p_max"  value="0.90"/>
    <param name="sdf_map/p_occ"  value="0.80"/>
    <param name="sdf_map/min_ray_length" value="0.5"/>
    <param name="sdf_map/max_ray_length" value="4.5"/>
    <param name="sdf_map/virtual_ceil_height"   value="-10"/>
    <param name="sdf_map/optimistic" value="true" type="bool"/>
    <param name="sdf_map/signed_dist" value="false" type="bool"/>
  
    <!--
    <param name="sdf_map/box_min_x" value="$(arg box_min_x)" type="double"/>
    <param name="sdf_map/box_min_y" value="$(arg box_min_y)" type="double"/>
    <param name="sdf_map/box_min_z" value="$(arg box_min_z)" type="double"/> 
    <param name="sdf_map/box_max_x" value="$(arg box_max_x)" type="double"/>
    <param name="sdf_map/box_max_y" value="$(arg box_max_y)" type="double"/>
    <param name="sdf_map/box_max_z" value="$(arg box_max_z)" type="double"/>
    -->
    <rosparam command="load" file="$(find exploration_manager)/config/maps/$(arg model).yaml" />
    
    <param name="sdf_map/no_drone_1" value="false" type="bool"/>

    <param name="map_ros/cx" value="$(arg cx)"/>
    <param name="map_ros/cy" value="$(arg cy)"/>
    <param name="map_ros/fx" value="$(arg fx)"/>
    <param name="map_ros/fy" value="$(arg fy)"/>
    <param name="map_ros/depth_filter_maxdist"   value="4.6"/>
    <param name="map_ros/depth_filter_mindist"   value="0.2"/>
    <param name="map_ros/depth_filter_margin"    value="2"/>
    <param name="map_ros/k_depth_scaling_factor" value="1000.0"/>
    <param name="map_ros/skip_pixel" value="2"/>
    <param name="map_ros/esdf_slice_height" value="0.5"/>
    <param name="map_ros/visualization_truncate_height"   value="10.09"/>
    <param name="map_ros/visualization_truncate_low"   value="-2.0"/>
    <param name="map_ros/show_occ_time"  value="false"/>
    <param name="map_ros/show_esdf_time" value="false"/>
    <param name="map_ros/show_all_map" value="true"/>
    <param name="map_ros/frame_id"      value="world"/>
    
    <!-- Logging -->
    <param name="multi_map_manager/log_folder" value="$(arg log_folder)"/>

    <!-- Fsm -->
    <param name="fsm/thresh_replan1" value="0.2" type="double"/>
    <param name="fsm/thresh_replan2" value="0.2" type="double"/>
    <param name="fsm/thresh_replan3" value="1.5" type="double"/>
    <!-- <param name="fsm/thresh_replan1" value="0.00001" type="double"/> -->
    <!-- <param name="fsm/thresh_replan2" value="100.0" type="double"/> -->
    <!-- <param name="fsm/thresh_replan3" value="0.1" type="double"/> -->
    <param name="fsm/replan_time" value="0.200" type="double"/>
    <param name="fsm/sync_interval" value="0.200" type="double"/>
    <param name="fsm/wait_delete_duration" value="0.05" type="double"/>
    <param name="fsm/gain_thresh" value="10" type="int"/>
    <param name="fsm/communication_range" value="$(arg communication_range)" type="double"/>

    <param name="fsm/attempt_interval" value="0.1" type="double"/>
    <param name="fsm/pair_opt_interval" value="0.5" type="double"/>

    <param name="partitioning/min_unknown" value="4000" type="int"/>;
    <param name="partitioning/min_frontier" value="100" type="int"/>;
    <param name="partitioning/min_free" value="3000" type="int"/>;
    <param name="partitioning/consistent_cost" value="-5" type="double"/>;
    <param name="partitioning/consistent_cost2" value="8" type="double"/>;
    <param name="partitioning/w_unknown" value="0.0" type="double"/>;
    <param name="partitioning/grid_size" value="15.0" type="double"/>;
    <param name="partitioning/use_swarm_tf" value="true" type="bool"/>;

    <!-- Exploration manager -->

    <param name="exploration/refine_local" value="true" type="bool"/>
    <param name="exploration/refined_num" value="7" type="int"/>
    <param name="exploration/refined_radius" value="5.0" type="double"/>
    <param name="exploration/max_decay" value="0.8" type="double"/>
    <param name="exploration/top_view_num" value="15" type="int"/>
    <param name="exploration/vm" value="$(eval 1.0 * arg('max_vel'))" type="double"/>
    <param name="exploration/am" value="$(eval 1.0 * arg('max_acc'))" type="double"/>
    <param name="exploration/yd" value="$(eval 80 * 3.1415926 / 180.0)" type="double"/>
    <!-- <param name="exploration/yd" value="$(eval 60 * 3.1415926 / 180.0)" type="double"/> -->
    <param name="exploration/ydd" value="$(eval 90 * 3.1415926 / 180.0)" type="double"/>
    <param name="exploration/w_dir" value="1.5" type="double"/>
    <param name="exploration/tsp_dir" value="$(find lkh_tsp_solver)/resource" type="string"/>
    <param name="exploration/mtsp_dir" value="$(find lkh_mtsp_solver)/resource" type="string"/>
    <param name="exploration/drone_num" value="$(arg drone_num)" type="double"/>
    <param name="exploration/drone_id" value="$(arg drone_id)" type="int"/>
    <param name="exploration/init_plan_num" value="2" type="int"/>
    
    <param name="frontier/cluster_min" value="100" type="int"/>
    <param name="frontier/cluster_size_xy" value="2.0" type="double"/>
    <param name="frontier/cluster_size_z" value="10.0" type="double"/>
    <param name="frontier/min_candidate_dist" value="0.5" type="double"/>
    <param name="frontier/min_candidate_clearance" value="0.41" type="double"/>
    <param name="frontier/candidate_dphi" value="$(eval 15 * 3.1415926 / 180.0)" type="double"/>
    <param name="frontier/candidate_rnum" value="3" type="int"/>
    <param name="frontier/candidate_rmin" value="1.0" type="double"/>
    <param name="frontier/candidate_rmax" value="1.5" type="double"/>
    <param name="frontier/down_sample" value="1" type="int"/>;
    <param name="frontier/min_visib_num" value="30" type="int"/>;
    <param name="frontier/min_view_finish_fraction" value="0.2" type="double"/>;


    <!-- Perception utils -->
    <param name="perception_utils/top_angle" value="0.56125" type="double"/>
    <param name="perception_utils/left_angle" value="0.69222" type="double"/>
    <param name="perception_utils/right_angle" value="0.68901" type="double"/>
    <param name="perception_utils/max_dist" value="4.5" type="double"/>
    <param name="perception_utils/vis_dist" value="1.0" type="double"/>


    <param name="heading_planner/yaw_diff" value="$(eval 30 * 3.1415926 / 180.0)" type="double"/>
    <param name="heading_planner/half_vert_num" value="5" type="int"/>
    <param name="heading_planner/lambda1" value="2.0" type="double"/>
    <param name="heading_planner/lambda2" value="1.0" type="double"/>
    <param name="heading_planner/max_yaw_rate" value="$(eval 10 * 3.1415926 / 180.0)" type="double"/>
    <param name="heading_planner/w" value="20000.0" type="double"/>
    <param name="heading_planner/weight_type" value="1" type="double"/>
    
    <!-- Role Assigner -->
    <param name="role_assigner/region_size" value="8.0" type="double"/>
    <param name="role_assigner/cluster_xy_size" value="3.0" type="double"/>
    <param name="role_assigner/min_num_neighbours" value="3" type="int"/>
    <param name="role_assigner/min_num_islands" value="4" type="int"/>
    <param name="role_assigner/min_num_trails" value="1" type="int"/>
    <!-- Ablation study -->
    <param name="role_assigner/fixed" value="$(arg role_assigner_fixed)" type="bool"/>
    <param name="role_assigner/fix_role" value="$(arg role_assigner_fix_role)" type="int"/>

    <!-- Collaboration Assigner -->
    <param name="collab_assigner/dist_range" value="15.0" type="double"/>
    <param name="collab_assigner/dist_collision" value="5.0" type="double"/>
    <param name="collab_assigner/w_range" value="0.7" type="double"/>
    <param name="collab_assigner/w_collision" value="0.3" type="double"/>
    <param name="collab_assigner/active" value="$(arg active_collaboration)" type="bool"/>
    
    <!-- Explorer parameters -->
    <param name="explorer/ftr_max_distance" value="6.0" type="double"/>
    <param name="explorer/max_ang_dist" value="$(eval 30.0 * 3.1415926 / 180.0)" type="double"/>
    <param name="explorer/label_penalty" value="0.0" type="double"/>
    <param name="explorer/w_distance" value="0.0" type="double"/>
    <param name="explorer/w_direction" value="10.0" type="double"/>
    <param name="explorer/w_others" value="1.0" type="double"/>
    <param name="explorer/w_previous_goal" value="1.0" type="double"/>
    
    <!-- Collector parameters -->
    <param name="collector/min_dist_collision" value="5.0" type="double"/>
    <param name="collector/min_vel" value="0.7" type="double"/>
    <param name="collector/label_penalty" value="15.0" type="double"/>
    <param name="collector/velocity_factor" value="2.0" type="double"/>
    <param name="collector/w_distance" value="3.0" type="double"/>
    <param name="collector/w_direction" value="1.0" type="double"/>
    <param name="collector/w_others" value="1.0" type="double"/>
    <param name="collector/w_previous_goal" value="3.0" type="double"/>
    
    <!-- Potential Field -->
    <param name="potential_field/ka" value="1.0" type="double"/>
    <param name="potential_field/kr" value="1.0" type="double"/>
    <param name="potential_field/d0" value="5.0" type="double"/>
    <param name="potential_field/df" value="6.0" type="double"/>
    <param name="potential_field/dc" value="2.0" type="double"/>

  <!-- planner manager -->
    <!-- <param name="manager/max_vel" value="$(arg max_vel)" type="double"/> -->
    <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/dynamic_environment" value="0" type="int"/>
    <param name="manager/local_segment_length" value="6.0" type="double"/>
    <param name="manager/clearance_threshold" value="0.2" type="double"/>
    <param name="manager/control_points_distance" value="0.5" type="double"/>
    <param name="manager/use_geometric_path" value="true" type="bool"/>
    <param name="manager/use_kinodynamic_path" value="true" type="bool"/>
    <param name="manager/use_topo_path" value="false" type="bool"/>
    <param name="manager/use_optimization" value="true" type="bool"/>
    <param name="manager/use_active_perception" value="true" type="bool"/>
    <param name="manager/min_time" value="true" type="bool"/>
    <param name="manager/relax_time1" value="0.3" type="double"/>
    <param name="manager/relax_time2" value="1.5" type="double"/>
    <param name="manager/max_yawdot" value="$(eval 120 * 3.1415926 / 180.0)" type="double"/>

  <!-- kinodynamic path searching -->
    <param name="search/max_tau" value="0.8" type="double"/>
    <param name="search/init_max_tau" value="1.0" type="double"/>
    <param name="search/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="search/vel_margin" value="0.25" type="double"/>
    <param name="search/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="search/w_time" value="10.0" type="double"/>
    <param name="search/horizon" value="5.0" type="double"/>
    <param name="search/lambda_heu" value="10.0" type="double"/>
    <param name="search/resolution_astar" value="0.025" type="double"/>
    <param name="search/time_resolution" value="0.8" type="double"/>
    <param name="search/margin" value="0.2" type="double"/>
    <param name="search/allocate_num" value="100000" type="int"/>
    <param name="search/check_num" value="10" type="int"/>
    <param name="search/optimistic" value="false" type="bool"/>

    <param name="astar/lambda_heu" value="10000.0" type="double"/>
    <param name="astar/resolution_astar" value="0.3" type="double"/>
    <param name="astar/allocate_num" value="1000000" type="int"/>
    <param name="astar/max_search_time" value="0.001" type="double"/>

  <!-- trajectory optimization -->
    <param name="optimization/ld_smooth" value="5.0" type="double"/>
    <param name="optimization/ld_dist" value="10.0" type="double"/>
    <param name="optimization/ld_feasi" value="2.0" type="double"/>
    <param name="optimization/ld_start" value="100.0" type="double"/>
    <param name="optimization/ld_end" value="0.5" type="double"/>
    <param name="optimization/ld_guide" value="1.5" type="double"/>
    <param name="optimization/ld_waypt" value="7.0" type="double"/>
    <param name="optimization/ld_view" value="0.0" type="double"/>
    <param name="optimization/ld_time" value="1.5" type="double"/>
    <param name="optimization/ld_swarm" value="5.0" type="double"/>
    <param name="optimization/swarm_safe_dist" value="1.0" type="double"/>
    <param name="optimization/dist0" value="0.7" type="double"/>
    <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="optimization/algorithm1" value="15" type="int"/>
    <param name="optimization/algorithm2" value="11" type="int"/>
    <param name="optimization/max_iteration_num1" value="2" type="int"/>
    <param name="optimization/max_iteration_num2" value="2000" type="int"/>
    <param name="optimization/max_iteration_num3" value="200" type="int"/>
    <param name="optimization/max_iteration_num4" value="200" type="int"/>

    <param name="optimization/max_iteration_time1" value="0.0001" type="double"/>
    <param name="optimization/max_iteration_time2" value="0.005" type="double"/>
    <param name="optimization/max_iteration_time3" value="0.003" type="double"/>
    <param name="optimization/max_iteration_time4" value="0.003" type="double"/>
    <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
    <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
    <param name="bspline/limit_ratio" value="1.1" type="double"/>
    <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
    <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
    <param name="bspline/limit_ratio" value="1.1" type="double"/>
  </node>

  <!-- <node pkg="lkh_tsp_solver" name="tsp_solver_$(arg drone_id)" type="tsp_node" output="screen">
    <param name="exploration/drone_id" value="$(arg drone_id)" type="int"/>
    <param name="exploration/tsp_dir"  value="$(find lkh_tsp_solver)/resource" type="string"/>
  </node> -->
  
  <node pkg="lkh_mtsp_solver" name="tsp_solver_$(arg drone_id)" type="mtsp_node" output="log">
    <param name="exploration/drone_id" value="$(arg drone_id)" type="int"/>
    <param name="exploration/mtsp_dir" value="$(find lkh_mtsp_solver)/resource" type="string"/>
    <param name="exploration/problem_id" value="1" type="int"/>
  </node>

  <node pkg="lkh_mtsp_solver" name="acvrp_solver_$(arg drone_id)" type="mtsp_node" output="log">
    <param name="exploration/drone_id" value="$(arg drone_id)" type="int"/>
    <param name="exploration/mtsp_dir" value="$(find lkh_mtsp_solver)/resource" type="string"/>
    <param name="exploration/problem_id" value="2" type="int"/>
  </node>
  
</launch>
